### PR TITLE
[SMTChecker] Fix SMT logic error when doing compound assignment with string literals

### LINKED ISSUE #10252: [SMTChecker] SMT logic error in function static solidity::smtutil::Expression solidity::smtutil::Expression::int2bv(solidity::smtutil::Expression, size_t)
## Description

This contract:
```solidity
pragma experimental SMTChecker;
contract C {
    function f3() public returns (bytes memory r) {
        bytes memory y = "def";
        y[0]&= "e";
    }
}
```

produces:

```
SMT logic error during analysis:
/home/user/solidity/libsmtutil/SolverInterface.h(212): Throw in function static solidity::smtutil::Expression solidity::smtutil::Expression::int2bv(solidity::smtutil::Expression, size_t)
Dynamic exception type: boost::exception_detail::clone_impl<solidity::smtutil::SMTLogicError>
std::exception::what: 
[solidity::util::tag_comment*] = 
```

when compiled with solc

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer. 

## Environment

- Compiler version:  Version: 0.7.5-develop.2020.11.10+commit.d3a016b5.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

### PR DESCRIPTION
**Pull request by mijovic**
Fixes #10252

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -1927,11 +1927,17 @@ smtutil::Expression SMTEncoder::compoundAssignment(Assignment const& _assignment
 
 	auto decl = identifierToVariable(_assignment.leftHandSide());
 
+	TypePointer commonType = Type::commonType(
+		_assignment.leftHandSide().annotation().type,
+		_assignment.rightHandSide().annotation().type
+	);
+	solAssert(commonType == _assignment.annotation().type, "");
```
Comment (mijovic) [inline libsolidity/formal/SMTEncoder.cpp:1934]: Maybe this assertion is not needed, but added it to be on a safe side.

**Review by hrkrshnn — DISMISSED** [review]

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1934]: I'd say you could even remove this block and just use `_assignment.annotation().type` in the modified `expr` calls

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Can you also add tests with strings of length >32?

**Comment by mijovic** [comment]
> Can you also add tests with strings of length >32?

Adding now

**Comment by mijovic** [comment]
> Can you also add tests with strings of length >32?

@leonardoalt to which type I can do compound ops with literals of length > 32. It fails compilation in type checker already on what I tried...

**Comment by leonardoalt** [comment]
@mijovic ok then let's keep it as is

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -1927,11 +1927,17 @@ smtutil::Expression SMTEncoder::compoundAssignment(Assignment const& _assignment
...
+
 	if (compoundToBitwise.count(op))
 		return bitwiseOperation(
 			compoundToBitwise.at(op),
-			decl ? currentValue(*decl) : expr(_assignment.leftHandSide()),
-			expr(_assignment.rightHandSide()),
+			decl ? currentValue(*decl) : expr(_assignment.leftHandSide(), _assignment.annotation().type),
+			expr(_assignment.rightHandSide(), _assignment.annotation().type),
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:1940]: @mijovic I think the two `expr` calls in `auto values = ...` should also be changed to be safe?

Comment (mijovic) [reply libsolidity/formal/SMTEncoder.cpp:1940]: True that. Will update them as well

Comment (mijovic) [reply libsolidity/formal/SMTEncoder.cpp:1940]: Fixed them

**Review by leonardoalt — DISMISSED** [review]

**Comment by leonardoalt** [comment]
Oh, it has conflicts

**Comment by mijovic** [comment]
> Oh, it has conflicts

Ah, changelog conflicts :)  Fixed now

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**